README.md
cloudscraper
============

Node.js library to bypass Cloudflare's anti-ddos page.

[![js-semistandard-style](https://cdn.rawgit.com/flet/semistandard/master/badge.svg)](https://github.com/Flet/semistandard)

[![Build status](https://img.shields.io/travis/codemanki/cloudscraper/master.svg?style=flat-square)](https://travis-ci.org/codemanki/cloudscraper)
[![Coverage](https://img.shields.io/coveralls/codemanki/cloudscraper.svg?style=flat-square)](https://coveralls.io/r/codemanki/cloudscraper)
[![Dependency Status](https://img.shields.io/david/codemanki/cloudscraper.svg?style=flat-square)](https://david-dm.org/codemanki/cloudscraper)
[![Greenkeeper badge](https://badges.greenkeeper.io/codemanki/cloudscraper.svg?style=flat-square)](https://greenkeeper.io/)

If the page you want to access is protected by Cloudflare, it will return special page, which expects client to support Javascript to solve challenge.

This small library encapsulates logic which extracts challenge, solves it, submits and returns the request page body.

You can use cloudscraper even if you are not sure if Cloudflare protection is turned on.

In general, Cloudflare has 4 types of _common_ anti-bot pages:
  - Simple html+javascript page with challenge
  - Page which redirects to original site
  - Page with reCAPTCHA
  - Page with error ( your ip was banned, etc)

If you notice that for some reason cloudscraper stops working, do not hesitate and get in touch with me ( by creating an issue [here](https://github.com/codemanki/cloudscraper/issues), for example), so i can update it.

Install
============
```sh
npm install cloudscraper
```

Saving the `request` module as a dependency is compulsory.

```sh
# Pin the request version
npm install --save request
```

Support for Brotli encoded responses is enabled by default when using Node.js v10 or later.
If you wish to enable support for older Node.js versions, you may install [brotli](https://npmjs.com/package/brotli).
It is recommended but not required.

Usage
============
Cloudscraper uses `request-promise` by default since v3. You can find the migration guide [here.](docs/migration-guide.md)

```javascript
var cloudscraper = require('cloudscraper');

cloudscraper.get('https://website.com/').then(console.log, console.error);
```

or for `POST` action:

```javascript
var options = {
  uri: 'https://website.com/',
  formData: { field1: 'value', field2: 2 }
};

cloudscraper.post(options).then(console.log).catch(console.error);
```

*Examples live in the docs directory of the Github repo and can be found [here.](docs/examples)*

A generic request can be made with `cloudscraper(options)`. The options object should follow [request's options](https://www.npmjs.com/package/request#request-options-callback). Not everything is supported however, for example http methods other than GET and POST. If you wanted to request an image in binary data you could use the encoding option:

```javascript
var options = {
  method: 'GET',
  url:'http://website.com/',
};

cloudscraper(options).then(console.log);
```

## Advanced usage
Cloudscraper allows you to specify your own requester, one of either `request` or `request-promise`.
Cloudscraper wraps the requester and accepts the same options, so using cloudscraper is pretty much like using those two libraries.
 - Cloudscraper exposes [the same HTTP verb methods as request](https://github.com/request/request#requestmethod):
   * `cloudscraper.get(options, callback)`
   * `cloudscraper.post(options, callback)`
   * `cloudscraper(uri)`
 - Cloudscraper uses request-promise by default, promise chaining is done exactly the same as described in [docs](https://github.com/request/request-promise#cheat-sheet):
 ```
  cloudscraper(options)
    .then(function (htmlString) {
    })
    .catch(function (err) {
    });
  ```
Please refer to the requester's documentation for further instructions.

## Sucuri
Cloudscraper can also identify and automatically bypass [Sucuri WAF](https://sucuri.net/website-firewall/). No actions are required.
 
## ReCAPTCHA
Cloudscraper may help you with the reCAPTCHA page. Take a look at [this example](docs/examples/solve-recaptcha.js) and an [example using promises](docs/examples/solve-recaptcha-v2.js).

Cloudflare may send a reCAPTCHA depending on the negotiated TLS cipher suite and extensions. Reducing the default cipher suite to only ciphers supported by Cloudflare may mitigate the problem: https://developers.cloudflare.com/ssl/ssl-tls/cipher-suites/

Only specifying the Cloudflare preferred TLSv1.2 cipher is also an option:
```javascript
var cloudscraper = require('cloudscraper').defaults({
  agentOptions: {
    ciphers: 'ECDHE-ECDSA-AES128-GCM-SHA256'
  }
})
```

More information on TLS issues can be found [here](https://github.com/codemanki/cloudscraper/issues?utf8=%E2%9C%93&q=tls).

## Defaults method

`cloudscraper.defaults` is a very convenient way of extending the cloudscraper requests with any of your settings.

```javascript
var cloudscraper = require('cloudscraper').defaults({ 'proxy': 'http://localproxy.com' });
// Overriding headers to remove them or using uncommon headers will cause reCAPTCHA responses
var headers = { /* ... */ };
var cloudscraper = require('cloudscraper').defaults({ headers: headers });

cloudscraper(options).then(console.log);
```

## Configuration
Cloudscraper exposes the following options that are required by default but might be changed. *Please note that the default values eliminate the chance of getting sent a CAPTCHA.*

```javascript
var options = {
  uri: 'https://website',
  jar: requestModule.jar(), // Custom cookie jar
  headers: {
    // User agent, Cache Control and Accept headers are required
    // User agent is populated by a random UA.
    'User-Agent': 'Ubuntu Chromium/34.0.1847.116 Chrome/34.0.1847.116 Safari/537.36',
    'Cache-Control': 'private',
    'Accept': 'application/xml,application/xhtml+xml,text/html;q=0.9, text/plain;q=0.8,image/png,*/*;q=0.5'
  },
  // Cloudscraper automatically parses out timeout required by Cloudflare.
  // Override cloudflareTimeout to adjust it.
  cloudflareTimeout: 5000,
  // Reduce Cloudflare's timeout to cloudflareMaxTimeout if it is excessive
  cloudflareMaxTimeout: 30000,
  // followAllRedirects - follow non-GET HTTP 3xx responses as redirects
  followAllRedirects: true,
  // Support only this max challenges in row. If CF returns more, throw an error
  challengesToSolve: 3,
  // Remove Cloudflare's email protection, replace encoded email with decoded versions
  decodeEmails: false,
  // Support gzip encoded responses (Should be enabled unless using custom headers)
  gzip: true,
  // Removes a few problematic TLSv1.0 ciphers to avoid CAPTCHA
  agentOptions: { ciphers }
};

cloudscraper(options).then(console.log);

```
You can access the default configuration with `cloudscraper.defaultParams`

## Error object
Cloudscraper error object inherits from `Error` has following fields:
  * `name` - `RequestError`/`CaptchaError`/`CloudflareError`/`ParserError`
  * `options` - The request options
  * `cause` - An alias for `error`
  * `response` - The request response
  * `errorType` - Custom error code
Where `errorType` can be following:
 - `0` if request to page failed due to some native reason as bad url, http connection or so. `error` in this case will be error [event](http://nodejs.org/api/http.html#http_class_http_server)
 - `1` Cloudflare returned CAPTCHA. Nothing to do here. Bad luck
 - `2` Cloudflare returned page with some inner error. `error` will be `Number` within this range `1012, 1011, 1002, 1000, 1004, 1010, 1006, 1007, 1008`. See more [here](https://support.cloudflare.com/hc/en-us/sections/200820298-Error-Pages)
 - `3` this error is returned when library failed to parse and solve js challenge. `error` will be `String` with some details. :warning: :warning: __Most likely it means that Cloudflare have changed their js challenge.__
 - `4` CF went into a loop and started to return challenge after challenge. If number of solved challenges is greater than `3` and another challenge is returned, throw an error

Errors are descriptive. You can find a list of all known errors [here.](errors.js)


Do not always rely on `error.cause` to be an error, it can be a string.

Running tests
============
Clone this repo, do `npm install` and then just `npm test`

### Unknown error? Library stopped working? ###
Let me know, by opening an [issue](https://github.com/codemanki/cloudscraper/issues) in this repo and I will update library asap. Please, provide url and body of page where cloudscraper failed.

WAT
===========
Current Cloudflare implementation requires browser to respect the timeout of 5 seconds and cloudscraper mimics this behaviour. So everytime you call `cloudscraper.get/post` you should expect it to return result after minimum 6 seconds. If you want to change this behaviour, you would need to make a generic request as described in above and pass `cloudflareTimeout` options with your value. But be aware that Cloudflare might track this timeout and use it against you ;)

## TODO
 - [x] Check for reCAPTCHA
 - [x] Support cookies, so challenge can be solved once per session
 - [x] Support page with simple redirects
 - [x] Add proper testing
 - [x] Remove manual 302 processing, replace with `followAllRedirects` param
 - [x] Parse out the timeout from challenge page
 - [x] Reorder the arguments in get/post/request methods and allow custom options to be passed in
 - [x] Support reCAPTCHA solving
 - [x] Promisification

## Kudos to contributors
 - [Dwayne](https://github.com/pro-src)
 - [drdokk](https://github.com/drdokk)
 - [Cole Faust](https://github.com/Colecf)
 - [Jeongbong Seo](https://github.com/jngbng)
 - [Mike van Rossum](https://github.com/askmike)
 - [Santiago Castro](https://github.com/bryant1410)
 - [Leonardo Gatica](https://github.com/lgaticaq)
 - [Michael](https://github.com/roflmuffin)
 - [Kamikadze4GAME](https://github.com/Kamikadze4GAME)
 - [Anorov](https://github.com/Anorov) :star:

In the beginning cloudscraper was a port of python module [cloudflare-scrape](https://github.com/Anorov/cloudflare-scrape). Thank you [Anorov](https://github.com/Anorov) for an inspiration.

## Dependencies
* [request-promise](https://github.com/request/request-promise)


CHANGELOG.md
## Change Log

### 4.6.0 (12/02/2020)
- Replace &amp; in url with `&`

### 4.5.0 (03/12/2019)
- [#293](https://github.com/codemanki/cloudscraper/pull/293) Update code to parse latest CF recaptcha.

### 4.4.0 (28/11/2019)
- [#288](https://github.com/codemanki/cloudscraper/pull/288) Update code to parse latest CF challenge.

### 4.3.0 (28/09/2019)
- [#267](https://github.com/codemanki/cloudscraper/pull/267) Typescript definitions.
- [#271](https://github.com/codemanki/cloudscraper/pull/271) Fix brotli compressed JSON responses.

### 4.2.0 (24/09/2019)
- [#260](https://github.com/codemanki/cloudscraper/pull/260) Update reCaptcha handling. Deprecate `captcha.url` in preference of `captcha.uri`. [Fix fallback siteKey handling](https://github.com/codemanki/cloudscraper/issues/259#issuecomment-531450844)

### 4.1.4 (24/08/2019)
- [#247](https://github.com/codemanki/cloudscraper/pull/247) Optimize header checks.

### 4.1.3 (12/07/2019)
- [#242](https://github.com/codemanki/cloudscraper/pull/242) Update Sucuri WAF Solving.

### 4.1.2 (23/05/2019)
- [#219](https://github.com/codemanki/cloudscraper/pull/219) Remove a few problematic TLSv1.0 ciphers.

### 4.1.1 (11/05/2019)
- Improve CF challenge security by nullifying VM context's prototype chain.

### v4.1.0 (02/05/2019)
- Backport TLSv1.3 secure ciphers to potentially avoid getting a CAPTCHA.

### v4.0.1 (25/04/2019)
- Improve documentation
- Add `url` to captcha
- Add more examples for reCAPTCHA handling

### v4.0.0 (22/04/2019)
- Randomize `User-Agent` header with random chrome browser
- Recaptcha solving support
- Brotli non-mandatory support
- Various code changes and improvements

### v3.9.1 (11/04/2019)
- Fix for the timeout parsing

### v3.9.0 (11/04/2019)
- [#193](https://github.com/codemanki/cloudscraper/pull/193) Fix bug with setTimeout match length

### v3.8.0 (11/04/2019)
- [#191](https://github.com/codemanki/cloudscraper/pull/191) Update code to parse latest CF challenge

### v3.7.0 (07/04/2019)
- [#182](https://github.com/codemanki/cloudscraper/pull/182) Usage examples have been added.
- [#169](https://github.com/codemanki/cloudscraper/pull/169) Cloudscraper now automatically parses out timeout for a CF challenge.

### v3.6.0 (03/04/2019)
- [#180](https://github.com/codemanki/cloudscraper/pull/180) Update code to parse latest CF challenge

### v3.5.0 (31/03/2019)
- [#174](https://github.com/codemanki/cloudscraper/pull/174) Update code to parse latest CF challenge

### v3.4.0 (27/03/2019)
- [#165](https://github.com/codemanki/cloudscraper/pull/165) Fixing CF challenge parsing, respect `Retry-After` header when CF returns `429 Too Many Requests` error.
- [#163](https://github.com/codemanki/cloudscraper/pull/163) Improve the accuracy of challenge timing. Throw error immediatelly without a delay
- [#159](https://github.com/codemanki/cloudscraper/pull/159) Decode emails in the page protected by CF

### v3.3.0 (22/03/2019)
- [#153](https://github.com/codemanki/cloudscraper/pull/153) Update code to parse latest CF challenge

### v3.2.0 (20/03/2019)
- [#149](https://github.com/codemanki/cloudscraper/pull/149) Update code to parse latest CF challenge

### v3.1.0 (14/03/2019)
- [#140](https://github.com/codemanki/cloudscraper/pull/140) Update code to parse new CF challenge

### v3.0.1 (11/03/2019)
- [#135](https://github.com/codemanki/cloudscraper/pull/135) Handle non-challenge response bodies 
- [#127](https://github.com/codemanki/cloudscraper/pull/127) Improve cloudflare detection 
- [#137](https://github.com/codemanki/cloudscraper/pull/137) Handle baseUrl option
- Various code style improvements

### v3.0.0 (07/03/2019)
- **BREAKING CHANGE**: `get/post` methods together with their signatures are aligned with corresponding methods from [request](https://github.com/request/request#requestmethod)
- **BREAKING CHANGE**: `cloudscraper.request` method is deprecated in favour of `cloudscraper(options)`
- Promise support has been added by using `request-promise`
- Error object are  inherited from Error and have additional properties.
  * `options` - The request options
  * `cause` - An alias for `error`
  * `response` - The request response
-  Stacktraces are available in error objects
- `cloudflareTimeout` option can be defined to speed up waiting time
- Challenge evaluation is done in a sandbox to avoid potential secutiry issues
- Default [request methods](https://github.com/request/request#requestmethod) are available
- Custom cookie jar can now be passed [#103](https://github.com/codemanki/cloudscraper/issues/102)
- Proxies support [PR#101](https://github.com/codemanki/cloudscraper/pull/101)
- MIT license

### v2.0.1 (02/03/2019)
- Minor documentation changes

### v2.0.0 (09/12/2018)
- [#2943](https://github.com/codemanki/cloudscraper/pull/66) Support recursive challenge solving. 
- **BREAKING CHANGE** Before this, when any error has been detected, the callback was called with an incorrect order: `callback(.., body, response);` instead of `return callback(..., response, body);`



errors.js
'use strict';

// The purpose of this library:
// 1. Have errors consistent with request/promise-core
// 2. Prevent request/promise core from wrapping our errors
// 3. Create descriptive errors.

// There are two differences between these errors and the originals.
// 1. There is a non-enumerable errorType attribute.
// 2. The error constructor is hidden from the stacktrace.

const EOL = require('os').EOL;
const original = require('request-promise-core/errors');
const http = require('http');

const BUG_REPORT = format([
  '### Cloudflare may have changed their technique, or there may be a bug.',
  '### Bug Reports: https://github.com/codemanki/cloudscraper/issues',
  '### Check the detailed exception message that follows for the cause.'
]);

const ERROR_CODES = {
  // Non-standard 5xx server error HTTP status codes
  520: 'Web server is returning an unknown error',
  521: 'Web server is down',
  522: 'Connection timed out',
  523: 'Origin is unreachable',
  524: 'A timeout occurred',
  525: 'SSL handshake failed',
  526: 'Invalid SSL certificate',
  527: 'Railgun Listener to Origin Error',
  530: 'Origin DNS error',
  // Other codes
  1000: 'DNS points to prohibited IP',
  1001: 'DNS resolution error',
  1002: 'Restricted or DNS points to Prohibited IP',
  1003: 'Access Denied: Direct IP Access Not Allowed',
  1004: 'Host Not Configured to Serve Web Traffic',
  1005: 'Access Denied: IP of banned ASN/ISP',
  1010: 'The owner of this website has banned your access based on your browser\'s signature',
  1011: 'Access Denied (Hotlinking Denied)',
  1012: 'Access Denied',
  1013: 'HTTP hostname and TLS SNI hostname mismatch',
  1016: 'Origin DNS error',
  1018: 'Domain is misconfigured',
  1020: 'Access Denied (Custom Firewall Rules)'
};

ERROR_CODES[1006] =
    ERROR_CODES[1007] =
        ERROR_CODES[1008] = 'Access Denied: Your IP address has been banned';

const OriginalError = original.RequestError;

const RequestError = create('RequestError', 0);
const CaptchaError = create('CaptchaError', 1);

// errorType 4 is a CloudflareError so this constructor is reused.
const CloudflareError = create('CloudflareError', 2, function (error) {
  if (!isNaN(error.cause)) {
    const description = ERROR_CODES[error.cause] || http.STATUS_CODES[error.cause];
    if (description) {
      error.message = error.cause + ', ' + description;
    }
  }
});

const ParserError = create('ParserError', 3, function (error) {
  error.message = BUG_REPORT + error.message;
});

// The following errors originate from promise-core and it's dependents.
// Give them an errorType for consistency.
original.StatusCodeError.prototype.errorType = 5;
original.TransformError.prototype.errorType = 6;

// This replaces the RequestError for all libraries using request/promise-core
// and prevents silent failure.
Object.defineProperty(original, 'RequestError', {
  configurable: true,
  enumerable: true,
  writable: true,
  value: RequestError
});

// Export our custom errors along with StatusCodeError, etc.
Object.assign(module.exports, original, {
  RequestError: RequestError,
  CaptchaError: CaptchaError,
  ParserError: ParserError,
  CloudflareError: CloudflareError
});

const desc = { configurable: true, writable: true, enumerable: false };
const descriptors = {
  error: desc,
  cause: desc,
  response: desc,
  options: desc
};

function create (name, errorType, customize) {
  function CustomError (cause, options, response) {
    // This prevents nasty things e.g. `error.cause.error` and
    // is why replacing the original RequestError is necessary.
    if (cause instanceof OriginalError) {
      return cause;
    }

    // Cleanup error output
    Object.defineProperties(this, descriptors);

    OriginalError.apply(this, arguments);

    // Change the name to match this constructor
    this.name = name;

    if (typeof customize === 'function') {
      customize(this);
    }

    if (Error.captureStackTrace) { // required for non-V8 environments
      // Provide a proper stack trace that hides this constructor
      Error.captureStackTrace(this, CustomError);
    }
  }

  CustomError.prototype = Object.create(OriginalError.prototype);
  CustomError.prototype.constructor = CustomError;
  // Keeps things stealthy by defining errorType on the prototype.
  // This makes it non-enumerable and safer to add.
  CustomError.prototype.errorType = errorType;

  Object.setPrototypeOf(CustomError, Object.getPrototypeOf(OriginalError));
  Object.defineProperty(CustomError, 'name', {
    configurable: true,
    value: name
  });

  return CustomError;
}

function format (lines) {
  return EOL + lines.join(EOL) + EOL + EOL;
}
